// get-db9/dist/client.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/client.ts
var client_exports = {};
__export(client_exports, {
  createDb9Client: () => createDb9Client
});
module.exports = __toCommonJS(client_exports);

// src/errors.ts
var Db9Error = class _Db9Error extends Error {
  statusCode;
  response;
  constructor(message, statusCode, response) {
    super(message);
    this.name = "Db9Error";
    this.statusCode = statusCode;
    this.response = response;
  }
  static async fromResponse(response) {
    let message;
    try {
      const body = await response.json();
      message = body.message || body.detail || body.error_description || body.error || response.statusText;
    } catch {
      message = response.statusText;
    }
    switch (response.status) {
      case 401:
        return new Db9AuthError(message, response);
      case 404:
        return new Db9NotFoundError(message, response);
      case 409:
        return new Db9ConflictError(message, response);
      default:
        return new _Db9Error(message, response.status, response);
    }
  }
};
var Db9AuthError = class extends Db9Error {
  constructor(message, response) {
    super(message, 401, response);
    this.name = "Db9AuthError";
  }
};
var Db9NotFoundError = class extends Db9Error {
  constructor(message, response) {
    super(message, 404, response);
    this.name = "Db9NotFoundError";
  }
};
var Db9ConflictError = class extends Db9Error {
  constructor(message, response) {
    super(message, 409, response);
    this.name = "Db9ConflictError";
  }
};

// src/http.ts
function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
function createHttpClient(options) {
  const fetchFn = options.fetch ?? globalThis.fetch;
  const baseUrl = options.baseUrl.replace(/\/$/, "");
  async function request(method, path, body, params) {
    let url = `${baseUrl}${path}`;
    if (params) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(params)) {
        if (value !== void 0) searchParams.set(key, value);
      }
      const qs = searchParams.toString();
      if (qs) url += `?${qs}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...options.headers
    };
    const init = { method, headers: reqHeaders };
    if (body !== void 0) {
      init.body = JSON.stringify(body);
    }
    const maxAttempts = Math.min(options.maxRetries ?? 0, 3) + 1;
    const baseDelay = options.retryDelay ?? 1e3;
    let lastError;
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      let timeoutId;
      try {
        const fetchInit = { ...init };
        if (options.timeout) {
          const controller = new AbortController();
          fetchInit.signal = controller.signal;
          timeoutId = setTimeout(() => controller.abort(), options.timeout);
        }
        const response = await fetchFn(url, fetchInit);
        if (timeoutId) clearTimeout(timeoutId);
        if (!response.ok) {
          if (response.status >= 500 && attempt < maxAttempts - 1) {
            lastError = await Db9Error.fromResponse(response);
            await delay(baseDelay * Math.pow(2, attempt));
            continue;
          }
          throw await Db9Error.fromResponse(response);
        }
        if (response.status === 204) return void 0;
        return response.json();
      } catch (err) {
        if (timeoutId) clearTimeout(timeoutId);
        if (err instanceof TypeError && attempt < maxAttempts - 1) {
          lastError = err;
          await delay(baseDelay * Math.pow(2, attempt));
          continue;
        }
        throw err;
      }
    }
    throw lastError;
  }
  async function requestRaw(method, path, body, params, customHeaders) {
    let url = `${baseUrl}${path}`;
    if (params) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(params)) {
        if (value !== void 0) searchParams.set(key, value);
      }
      const qs = searchParams.toString();
      if (qs) url += `?${qs}`;
    }
    const headers = {
      ...options.headers,
      ...customHeaders
    };
    const fetchInit = { method, headers };
    if (body !== void 0) fetchInit.body = body;
    let timeoutId;
    if (options.timeout) {
      const controller = new AbortController();
      fetchInit.signal = controller.signal;
      timeoutId = setTimeout(() => controller.abort(), options.timeout);
    }
    try {
      const response = await fetchFn(url, fetchInit);
      if (timeoutId) clearTimeout(timeoutId);
      if (!response.ok) throw await Db9Error.fromResponse(response);
      return response;
    } catch (err) {
      if (timeoutId) clearTimeout(timeoutId);
      throw err;
    }
  }
  return {
    get: (path, params) => request("GET", path, void 0, params),
    post: (path, body) => request("POST", path, body),
    put: (path, body) => request("PUT", path, body),
    del: (path) => request("DELETE", path),
    getRaw: (path, params) => requestRaw("GET", path, void 0, params),
    putRaw: (path, body, headers) => requestRaw("PUT", path, body, void 0, headers),
    postRaw: (path, body, headers) => requestRaw("POST", path, body, void 0, headers),
    delRaw: (path, params) => requestRaw("DELETE", path, void 0, params)
  };
}

// src/credentials.ts
var import_toml = require("@iarna/toml");
var FileCredentialStore = class {
  customPath;
  /**
   * @param path — Override the credential file location.
   *               Defaults to `~/.db9/credentials` (resolved lazily).
   */
  constructor(path) {
    this.customPath = path;
  }
  /** Resolve the credential file path (lazy to avoid top-level `os` import). */
  async resolvePath() {
    if (this.customPath) return this.customPath;
    const os = await import("os");
    const nodePath = await import("path");
    return nodePath.join(os.homedir(), ".db9", "credentials");
  }
  async load() {
    const fs = await import("fs/promises");
    const filePath = await this.resolvePath();
    let content;
    try {
      content = await fs.readFile(filePath, "utf-8");
    } catch (err) {
      if (err.code === "ENOENT") return null;
      throw err;
    }
    const parsed = (0, import_toml.parse)(content);
    const token = parsed["token"];
    if (typeof token !== "string") return null;
    return { token };
  }
  async save(credentials) {
    const fs = await import("fs/promises");
    const nodePath = await import("path");
    const filePath = await this.resolvePath();
    const dir = nodePath.dirname(filePath);
    await fs.mkdir(dir, { recursive: true, mode: 448 });
    const data = {};
    try {
      const raw = await fs.readFile(filePath, "utf-8");
      const parsed = (0, import_toml.parse)(raw);
      for (const [k, v] of Object.entries(parsed)) {
        if (typeof v === "string" || typeof v === "boolean" || typeof v === "number") {
          data[k] = v;
        }
      }
    } catch (err) {
      if (err.code !== "ENOENT") throw err;
    }
    data["token"] = credentials.token;
    const toml = (0, import_toml.stringify)(
      data
    );
    await fs.writeFile(filePath, toml, { mode: 384 });
  }
  async clear() {
    const fs = await import("fs/promises");
    const filePath = await this.resolvePath();
    try {
      await fs.unlink(filePath);
    } catch (err) {
      if (err.code !== "ENOENT") throw err;
    }
  }
};
function defaultCredentialStore() {
  return new FileCredentialStore();
}

// src/ws.ts
var import_node_crypto = require("crypto");
var FsError = class extends Error {
  code;
  constructor(code, message) {
    super(message);
    this.name = "FsError";
    this.code = code;
  }
};
var STREAMING_THRESHOLD = 1024 * 1024;
var DEFAULT_CHUNK_SIZE = 64 * 1024;
var nextId = 1;
function nextRequestId() {
  return String(nextId++);
}
function toBase64(data) {
  const bytes = data instanceof Uint8Array ? data : new Uint8Array(data);
  return Buffer.from(bytes).toString("base64");
}
function fromBase64(b64) {
  return new Uint8Array(Buffer.from(b64, "base64"));
}
function computeChecksum(data) {
  const hash = (0, import_node_crypto.createHash)("sha256").update(data).digest("hex");
  return `sha256:${hash}`;
}
function encodeBinaryFrame(streamId, chunk) {
  const frame = Buffer.alloc(8 + chunk.length);
  frame.writeBigUInt64BE(BigInt(streamId), 0);
  frame.set(chunk, 8);
  return frame;
}
var FsClient = class _FsClient {
  ws;
  pending = /* @__PURE__ */ new Map();
  closed = false;
  constructor(ws) {
    this.ws = ws;
    ws.onmessage = (ev) => {
      const text = typeof ev.data === "string" ? ev.data : String(ev.data);
      let resp;
      try {
        resp = JSON.parse(text);
      } catch {
        return;
      }
      const p = this.pending.get(resp.id);
      if (p) {
        this.pending.delete(resp.id);
        p.resolve(resp);
      }
    };
    ws.onclose = () => {
      this.closed = true;
      for (const [, p] of this.pending) {
        p.reject(new FsError("CONNECTION_CLOSED", "WebSocket connection closed"));
      }
      this.pending.clear();
    };
    ws.onerror = () => {
    };
  }
  /**
   * Connect to an fs9 WebSocket server.
   *
   * @param url WebSocket URL, e.g. `wss://host:5480`
   * @param WS  WebSocket constructor (native or from `ws` package)
   */
  static connect(url, WS) {
    return new Promise((resolve, reject) => {
      const ws = new WS(url);
      ws.onopen = () => {
        ws.onopen = null;
        ws.onerror = null;
        resolve(new _FsClient(ws));
      };
      ws.onerror = (ev) => {
        ws.onopen = null;
        ws.onerror = null;
        const msg = ev && typeof ev === "object" && "message" in ev ? String(ev.message) : "WebSocket connection failed";
        reject(new FsError("CONNECTION_ERROR", msg));
      };
    });
  }
  /** Authenticate with the server. Must be called first after connect. */
  async authenticate(username, password) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "auth",
      username,
      password
    });
    if (!resp.ok) {
      throw new FsError("AUTH_FAILED", this.errorMessage(resp));
    }
    const data = resp.data;
    return {
      user: String(data?.user ?? ""),
      tenant: String(data?.tenant ?? ""),
      keyspace: String(data?.keyspace ?? "")
    };
  }
  /** Get file or directory metadata. */
  async stat(path) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "stat",
      path
    });
    this.expectOk(resp);
    return resp.data;
  }
  /** List directory contents. */
  async readdir(path) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "readdir",
      path
    });
    this.expectOk(resp);
    const data = resp.data;
    return data?.entries ?? [];
  }
  /** Create a directory. Always recursive (mkdir -p). */
  async mkdir(path, recursive = true) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "mkdir",
      path,
      recursive
    });
    this.expectOk(resp);
  }
  /** Read an entire file, returning raw bytes. */
  async readFile(path) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "read",
      path
    });
    this.expectOk(resp);
    const data = resp.data;
    const content = data?.content;
    if (typeof content !== "string") {
      throw new FsError("PROTOCOL", "missing content field in read response");
    }
    return fromBase64(content);
  }
  /**
   * Write (overwrite) a file. Returns bytes written.
   *
   * Automatically uses streaming mode for files >= 1 MB to avoid base64
   * overhead and bypass the 2 MB JSON frame limit.
   */
  async writeFile(path, data) {
    const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data instanceof Uint8Array ? data : new Uint8Array(data);
    if (bytes.byteLength >= STREAMING_THRESHOLD) {
      return this.writeFileStreaming(path, bytes);
    }
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "write",
      path,
      content: toBase64(bytes),
      encoding: "base64"
    });
    this.expectOk(resp);
    const result = resp.data;
    return result?.written ?? bytes.byteLength;
  }
  /** Append to a file. Returns bytes written. */
  async appendFile(path, data) {
    const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data instanceof Uint8Array ? data : new Uint8Array(data);
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "append",
      path,
      content: toBase64(bytes),
      encoding: "base64"
    });
    this.expectOk(resp);
    const result = resp.data;
    return result?.written ?? bytes.byteLength;
  }
  /** Remove a file (non-recursive) or directory (recursive). */
  async rm(path, recursive = false) {
    const resp = await this.sendAndRecv(
      recursive ? { id: nextRequestId(), op: "rm", path, recursive: true } : { id: nextRequestId(), op: "unlink", path }
    );
    this.expectOk(resp);
  }
  /** Rename (move) a file or directory. */
  async rename(oldPath, newPath) {
    const resp = await this.sendAndRecv({
      id: nextRequestId(),
      op: "rename",
      old_path: oldPath,
      new_path: newPath
    });
    this.expectOk(resp);
  }
  /** Gracefully close the WebSocket connection. */
  async close() {
    if (!this.closed) {
      this.ws.close();
    }
  }
  // ── streaming internals ────────────────────────────────────────
  /**
   * Write a file using streaming mode (binary frames, no base64).
   *
   * Protocol:
   * 1. Send streaming write request with file size
   * 2. Receive ready response with stream_id and chunk_size
   * 3. Send binary frames: [8-byte stream_id BE][chunk_data]
   * 4. Send stream end with checksum
   * 5. Receive final write confirmation
   */
  async writeFileStreaming(path, bytes) {
    const requestId = nextRequestId();
    const readyResp = await this.sendAndRecv({
      id: requestId,
      op: "write",
      path,
      streaming: true,
      size: bytes.byteLength
    });
    this.expectOk(readyResp);
    const ready = readyResp.data;
    if (!ready?.ready) {
      throw new FsError("PROTOCOL", "server not ready for streaming");
    }
    const streamId = ready.stream_id;
    const chunkSize = ready.chunk_size > 0 ? ready.chunk_size : DEFAULT_CHUNK_SIZE;
    try {
      for (let offset = 0; offset < bytes.byteLength; offset += chunkSize) {
        const chunk = bytes.subarray(offset, offset + chunkSize);
        const frame = encodeBinaryFrame(streamId, chunk);
        this.ws.send(frame);
      }
    } catch (err) {
      this.tryAbortStream(streamId);
      throw new FsError("CONNECTION_ERROR", `send chunk: ${err}`);
    }
    const checksum = computeChecksum(bytes);
    const finalResp = await this.sendAndRecv({
      id: requestId,
      stream: "end",
      stream_id: streamId,
      checksum
    });
    this.expectOk(finalResp);
    const result = finalResp.data;
    return result?.written ?? bytes.byteLength;
  }
  /** Best-effort abort of an in-progress stream so the server can clean up. */
  tryAbortStream(streamId) {
    try {
      this.ws.send(JSON.stringify({ stream: "abort", stream_id: streamId }));
    } catch {
    }
  }
  // ── internals ──────────────────────────────────────────────────
  sendAndRecv(request) {
    if (this.closed) {
      return Promise.reject(new FsError("CONNECTION_CLOSED", "WebSocket is closed"));
    }
    return new Promise((resolve, reject) => {
      const id = request.id;
      this.pending.set(id, { resolve, reject });
      try {
        this.ws.send(JSON.stringify(request));
      } catch (err) {
        this.pending.delete(id);
        reject(new FsError("CONNECTION_ERROR", String(err)));
      }
    });
  }
  expectOk(resp) {
    if (!resp.ok) {
      const detail = resp.error;
      throw new FsError(
        detail?.code ?? "UNKNOWN",
        detail?.message ?? "unknown error"
      );
    }
  }
  errorMessage(resp) {
    const detail = resp.error;
    return detail ? `${detail.code}: ${detail.message}` : "unknown error";
  }
};

// src/client.ts
function createDb9Client(options = {}) {
  const baseUrl = options.baseUrl ?? "https://db9.ai/api";
  let token = options.token;
  let tokenLoaded = !!token;
  const store = options.credentialStore ?? defaultCredentialStore();
  const fetchFn = options.fetch ?? globalThis.fetch;
  const publicClient = createHttpClient({
    baseUrl,
    fetch: options.fetch,
    timeout: options.timeout,
    maxRetries: options.maxRetries,
    retryDelay: options.retryDelay
  });
  async function getAuthClient() {
    if (!token && !tokenLoaded) {
      const creds = await store.load();
      if (creds?.token) token = creds.token;
      tokenLoaded = true;
    }
    if (!token) {
      throw new Db9Error(
        "No token available. Run `db9 login` or pass a token via Db9ClientOptions.",
        401
      );
    }
    return createHttpClient({
      baseUrl,
      fetch: options.fetch,
      headers: { Authorization: `Bearer ${token}` },
      timeout: options.timeout,
      maxRetries: options.maxRetries,
      retryDelay: options.retryDelay
    });
  }
  async function withAuthRetry(operation) {
    const client = await getAuthClient();
    return operation(client);
  }
  const fsWsPort = options.wsPort ?? 5480;
  async function resolveFsConn(dbId) {
    const creds = await withAuthRetry(
      (client) => client.get(
        `/customer/databases/${dbId}/credentials`
      )
    );
    const connStr = creds.connection_string;
    const hostMatch = connStr.match(/@([^:/?]+)/);
    const host = hostMatch?.[1];
    if (!host) {
      throw new Error(`Cannot parse host from connection string for database '${dbId}'`);
    }
    const userMatch = connStr.match(/:\/\/([^:@]+)/);
    const username = userMatch?.[1] ?? creds.admin_user;
    const protocol = host === "localhost" || host === "127.0.0.1" ? "ws" : "wss";
    return {
      wsUrl: `${protocol}://${host}:${fsWsPort}`,
      username,
      password: creds.admin_password
    };
  }
  async function withFsClient(dbId, operation) {
    const WS = options.WebSocket ?? (typeof globalThis !== "undefined" ? globalThis.WebSocket : void 0);
    if (!WS) {
      throw new Error(
        "WebSocket constructor not available. Pass `WebSocket` in Db9ClientOptions, or use Node 21+ / a browser environment with native WebSocket support, or install the `ws` package for Node 18\u201320."
      );
    }
    const conn = await resolveFsConn(dbId);
    const client = await FsClient.connect(conn.wsUrl, WS);
    try {
      await client.authenticate(conn.username, conn.password);
      return await operation(client);
    } finally {
      await client.close();
    }
  }
  function parseSqlError(raw) {
    try {
      const parsed = JSON.parse(raw);
      if (typeof parsed === "object" && parsed !== null && typeof parsed.message === "string") {
        return parsed;
      }
    } catch {
    }
    const pgMatch = raw.match(/^(?:ERROR:\s*)?(.+?)(?:\s+DETAIL:\s+(.+?))?(?:\s+HINT:\s+(.+?))?(?:\s+\(SQLSTATE\s+(\w+)\))?$/s);
    if (pgMatch && pgMatch[1]) {
      const result = { message: pgMatch[1].trim() };
      if (pgMatch[2]) result.detail = pgMatch[2].trim();
      if (pgMatch[3]) result.hint = pgMatch[3].trim();
      if (pgMatch[4]) result.code = pgMatch[4];
      return result;
    }
    return { message: raw };
  }
  return {
    auth: {
      me: async () => withAuthRetry(
        (client) => client.get("/customer/me")
      )
    },
    tokens: {
      list: async () => withAuthRetry(
        (client) => client.get("/customer/tokens")
      ),
      revoke: async (tokenId) => withAuthRetry(
        (client) => client.del(`/customer/tokens/${tokenId}`)
      ),
      create: async (req) => withAuthRetry(
        (client) => client.post("/customer/tokens", req)
      )
    },
    databases: {
      // ── CRUD ──────────────────────────────────────────────────
      create: async (req) => withAuthRetry(
        (client) => client.post("/customer/databases", req)
      ),
      list: async () => withAuthRetry(
        (client) => client.get("/customer/databases")
      ),
      get: async (databaseId) => withAuthRetry(
        (client) => client.get(
          `/customer/databases/${databaseId}`
        )
      ),
      delete: async (databaseId) => withAuthRetry(
        (client) => client.del(
          `/customer/databases/${databaseId}`
        )
      ),
      resetPassword: async (databaseId) => withAuthRetry(
        (client) => client.post(
          `/customer/databases/${databaseId}/reset-password`
        )
      ),
      credentials: async (databaseId) => withAuthRetry(
        (client) => client.get(
          `/customer/databases/${databaseId}/credentials`
        )
      ),
      observability: async (databaseId) => withAuthRetry(
        (client) => client.get(
          `/customer/databases/${databaseId}/observability`
        )
      ),
      // ── SQL Execution ─────────────────────────────────────────
      sql: async (databaseId, query) => {
        const result = await withAuthRetry(
          (client) => client.post(
            `/customer/databases/${databaseId}/sql`,
            { query }
          )
        );
        if (result.error && typeof result.error === "string") {
          result.error = parseSqlError(result.error);
        }
        return result;
      },
      sqlFile: async (databaseId, fileContent) => {
        const result = await withAuthRetry(
          (client) => client.post(
            `/customer/databases/${databaseId}/sql`,
            { file_content: fileContent }
          )
        );
        if (result.error && typeof result.error === "string") {
          result.error = parseSqlError(result.error);
        }
        return result;
      },
      // ── Schema & Dump ─────────────────────────────────────────
      schema: async (databaseId) => withAuthRetry(
        (client) => client.get(
          `/customer/databases/${databaseId}/schema`
        )
      ),
      dump: async (databaseId, req) => withAuthRetry(
        (client) => client.post(
          `/customer/databases/${databaseId}/dump`,
          req
        )
      ),
      // ── Migrations ────────────────────────────────────────────
      applyMigration: async (databaseId, req) => withAuthRetry(
        (client) => client.post(
          `/customer/databases/${databaseId}/migrations`,
          req
        )
      ),
      listMigrations: async (databaseId) => withAuthRetry(
        (client) => client.get(
          `/customer/databases/${databaseId}/migrations`
        )
      ),
      // ── Branching ─────────────────────────────────────────────
      branch: async (databaseId, req) => withAuthRetry(
        (client) => client.post(
          `/customer/databases/${databaseId}/branch`,
          req
        )
      ),
      // ── User Management ───────────────────────────────────────
      users: {
        list: async (databaseId) => withAuthRetry(
          (client) => client.get(
            `/customer/databases/${databaseId}/users`
          )
        ),
        create: async (databaseId, req) => withAuthRetry(
          (client) => client.post(
            `/customer/databases/${databaseId}/users`,
            req
          )
        ),
        delete: async (databaseId, username) => withAuthRetry(
          (client) => client.del(
            `/customer/databases/${databaseId}/users/${username}`
          )
        )
      }
    },
    fs: {
      /**
       * Open a persistent WebSocket connection for multiple fs operations.
       * Caller is responsible for calling `client.close()` when done.
       */
      connect: async (dbId) => {
        const WS = options.WebSocket ?? (typeof globalThis !== "undefined" ? globalThis.WebSocket : void 0);
        if (!WS) {
          throw new Error(
            "WebSocket constructor not available. Pass `WebSocket` in Db9ClientOptions, or use Node 21+ / a browser environment with native WebSocket support, or install the `ws` package for Node 18\u201320."
          );
        }
        const conn = await resolveFsConn(dbId);
        const client = await FsClient.connect(conn.wsUrl, WS);
        await client.authenticate(conn.username, conn.password);
        return client;
      },
      /** List directory contents. */
      list: async (dbId, path) => withFsClient(dbId, (client) => client.readdir(path)),
      /** Read a file as text (UTF-8). */
      read: async (dbId, path) => withFsClient(dbId, async (client) => {
        const bytes = await client.readFile(path);
        return new TextDecoder().decode(bytes);
      }),
      /** Read a file as raw bytes. */
      readBinary: async (dbId, path) => withFsClient(dbId, (client) => client.readFile(path)),
      /** Write (overwrite) a file. Accepts string, ArrayBuffer, or Uint8Array. */
      write: async (dbId, path, content) => {
        await withFsClient(dbId, (client) => client.writeFile(path, content));
      },
      /** Append to a file. Returns bytes written. */
      append: async (dbId, path, content) => withFsClient(dbId, (client) => client.appendFile(path, content)),
      /** Get file or directory metadata. */
      stat: async (dbId, path) => withFsClient(dbId, (client) => client.stat(path)),
      /** Check if a file or directory exists. */
      exists: async (dbId, path) => {
        try {
          await withFsClient(dbId, (client) => client.stat(path));
          return true;
        } catch (err) {
          if (err instanceof FsError && err.code === "ENOENT") {
            return false;
          }
          throw err;
        }
      },
      /** Create a directory (recursive by default). */
      mkdir: async (dbId, path) => {
        await withFsClient(dbId, (client) => client.mkdir(path, true));
      },
      /** Remove a file or directory. */
      remove: async (dbId, path, opts) => {
        await withFsClient(dbId, (client) => client.rm(path, opts?.recursive ?? false));
      },
      /** Rename (move) a file or directory. */
      rename: async (dbId, oldPath, newPath) => {
        await withFsClient(dbId, (client) => client.rename(oldPath, newPath));
      }
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createDb9Client
});
